airbyte-config/persistence/src/main/java/io/airbyte/config/persistence/ConfigRepository.java

```diff
@@ -615,35 +629,86 @@ public void updateConnectionState(final UUID connectionId, final State state) th
     return actorCatalogs;
   }
 
-  public void writeCatalog(final AirbyteCatalog catalog,
-                           final UUID sourceId,
-                           final String configurationHash,
-                           final String connectorVersion)
-      throws JsonValidationException, IOException {
+  public Map<UUID, AirbyteCatalog> findCatalogByHash(final String catalogHash) throws IOException {
+    final Result<Record2<UUID, JSONB>> records = database.transaction(ctx -> ctx.select(ACTOR_CATALOG.ID, ACTOR_CATALOG.CATALOG)
+        .from(ACTOR_CATALOG)
+        .where(ACTOR_CATALOG.CATALOG_HASH.eq(catalogHash))).fetch();
+
+    final Map<UUID, AirbyteCatalog> result = new HashMap<>();
+    for (final Record record : records) {
+      final AirbyteCatalog catalog = Jsons.deserialize(
+          record.get(ACTOR_CATALOG.CATALOG).toString(), AirbyteCatalog.class);
+      result.put(record.get(ACTOR_CATALOG.ID), catalog);
+    }
+    return result;
+  }
+
+  public Optional<AirbyteCatalog> getActorCatalog(final UUID actorId,
+                                                  final String actorVersion,
+                                                  final String configHash)
+      throws IOException {
+    final Result<Record1<JSONB>> records = database.transaction(ctx -> ctx.select(ACTOR_CATALOG.CATALOG)
+        .from(ACTOR_CATALOG).join(ACTOR_CATALOG_FETCH_EVENT)
+        .on(ACTOR_CATALOG.ID.eq(ACTOR_CATALOG_FETCH_EVENT.ACTOR_CATALOG_ID))
+        .where(ACTOR_CATALOG_FETCH_EVENT.ACTOR_ID.eq(actorId))
+        .and(ACTOR_CATALOG_FETCH_EVENT.ACTOR_VERSION.eq(actorVersion))
+        .and(ACTOR_CATALOG_FETCH_EVENT.CONFIG_HASH.eq(configHash))).fetch();
+
+    if (records.size() > 1) {
+      throw new IllegalStateException(
+          String.format("Multiple catalog stored for ActorId=%s,ActorVersion=%s,ConfigHash=%s", actorId.toString(), actorVersion, configHash));
```

Comment:
In the situation you described Charles, wouldn't the config hash be different if the catalog has changed? So this likely wouldn't return anything right? I may be misunderstanding something here

Reply:
I don't think so. Let's take an example. Let's say for postgres my config is just username: charles, password: 123. If call discover multiple times in a row, sometimes I might get back the same catalog (the schema can change even though the config didnt). It is also possible that I'll get back the _same_ catalog. This will happen in the case where the schema hasn't change but the user has forced a discover call to run anyway. In which case there will be multiple fetch events pointing at the same catalog.

I think in the case where there are multiple we should use whatever the last fetched catalog is. Does that make sense?

Reply:
i think if the query adds an order by fetch_event.created_at and does a limit 1 then we can keep this check.

Reply:
@malikdiarra lmk if there's still an open question on this.

Reply:
Yes, that the remaining change I have to do to that review. I'll remove the exception being raised and will instead:
- order by created_at desc
- return the first event I am getting

Reply:
Fixed, I'm ordering by creation date right and doing a limit 1. I removed that check entirely as having more than one row in the results in such would constitute a database bug.